contracts/src/FunctionGatewayV2.sol

```diff
@@ -0,0 +1,165 @@
+contract FunctionGatewayUpdated {
+    mapping(bytes32 => address) functionIdToVerifier;
+
+    function callbackRequestV1(
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes4 _callbackSelector,
+        bytes memory _context,
+        uint8 _preference // Slow, medium, fast, aggregation or not, etc.
+    ) external payable returns (bytes32) {
+        requests[nonce++] = keccak256(FunctionRequest({
+            functionId: _functionId,
+            inputBytes: _input,
+            context: _context,
+            callbackAddress: msg.sender,
+            callbackSelector: _callbackSelector
+        }));
+        // TODO: emit event with preference
+    }
+
+    function fulfillCallbackRequestV1(
+        bytes32 _requestId,
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _output,
+        bytes memory _proof,
+        bytes memory _context,
+        address _callbackAddress,
+        bytes32 _callbackSelector
+    ) external {
+        bytes32 requestHash = keccak256(FunctionRequest({
+            functionId: _functionId,
+            inputBytes: _input,
+            context: _context,
+            callbackAddress: _callbackAddress,
+            callbackSelector: _callbackSelector
+        }));
+        require(requests[_requestId] == requestHash, "Request not found");
+
+        address verifier = functionIdToVerifier[_functionId];
+        bytes32 inputHash = sha256(_input);
+        bytes32 outputHash = sha256(_output);
+        if (
+            !IFunctionVerifier(verifier).verify(
+                inputHash,
+                outputHash,
+                _proof
+            )
+        ) {
+            revert InvalidProof();
+        }
+
+        (bool status, ) = _callbackAddress.call(
+            abi.encodeWithSelector(_callbackSelector, _input, _output, _context)
+        );
+        if (!status) {
+            revert CallbackFailed(_callbackAddress, _callbackSelector);
+        }
+
+        delete requests[_requestId];
```

Comment:
Not relevant to the changes but please remember to `delete requests[_requestId];` before the .call